src/App.jsx

```diff
@@ -1,248 +1,333 @@
+  };
+});
+
+function filterProducts(productsTable, filterByUser, filterByCategory, query) {
+  let filteredProducts = [...productsTable];
+
+  if (query) {
+    const lowerQuery = query.toLowerCase();
+
+    filteredProducts = productsTable
+      .filter(product => product.name.toLowerCase().includes(lowerQuery));
+  }
+
+  if (filterByUser !== 'all') {
+    filteredProducts = filteredProducts
+      .filter(product => product.user.id === filterByUser);
+  }
+
+  if (filterByCategory.length !== 0) {
+    filteredProducts = filteredProducts
+      .filter(product => filterByCategory.includes(product.categoryId));
+  }
+
+  return filteredProducts;
+}
+
+function sortProducts(visibleProducts, sortByColumn, isSortReversed) {
+  const sortedProducts = [...visibleProducts];
+
+  if (sortByColumn) {
+    sortedProducts.sort((product1, product2) => {
+      switch (sortByColumn) {
+        case 'id':
+          return product1.id - product2.id;
+        case 'product':
+          return product1.name.localeCompare(product2.name);
+        case 'category':
+          return product1.category.title.localeCompare(product2.category.title);
+        case 'user':
+          return product1.user.name.localeCompare(product2.user.name);
+        default: return 0;
+      }
+    });
+  }
+
+  return isSortReversed ? sortedProducts.reverse() : sortedProducts;
+}
+
+export const App = () => {
+  const [filterByUser, setFilterByUser] = useState('all');
+  const [filterByCategories, setFilterByCategories] = useState([]);
+  const [query, setQuery] = useState('');
+  const [sortByColumn, setSortByColumn] = useState('');
+  const [isSortReversed, setIsSortReversed] = useState(false);
+
+  const visibleProducts
+    = filterProducts(products, filterByUser, filterByCategories, query);
+
+  const productsToRender
+    = sortProducts(visibleProducts, sortByColumn, isSortReversed);
+
+  const handleCategoriesClick = (categId) => {
+    if (filterByCategories.includes(categId)) {
+      setFilterByCategories(filterByCategories.filter(el => el !== categId));
+    } else {
+      setFilterByCategories([
+        ...filterByCategories,
+        categId,
+      ]);
+    }
+  };
+
+  const handleSortClick = (sortBy) => {
+    if (sortByColumn === sortBy) {
+      setIsSortReversed(true);
+    }
+
+    if (isSortReversed) {
+      setSortByColumn('');
+      setIsSortReversed(false);
+    }
+
+    setSortByColumn(sortBy);
+  };
+
+  return (
+    <div className="section">
+      <div className="container">
+        <h1 className="title">Product Categories</h1>
+
+        <div className="block">
+          <nav className="panel">
+            <p className="panel-heading">Filters</p>
+
+            <p className="panel-tabs has-text-weight-bold">
+              <a
+                data-cy="FilterAllUsers"
+                href="#/"
+                className={cn({ 'is-active': filterByUser === 'all' })}
+                onClick={() => setFilterByUser('all')}
```

Comment:
```suggestion
                onClick={() => setFilterByUser('')}
```
Actually you don't have to add some specific order in this case